Add Dynamic Process Title Support

mattsta opened the pull request:
This implements a dynamically updating process title based on the state of the Redis server.

The format of the new process title is:

```
{{executable}} {{name}} {{config}} {{ip0}}:{{port}} {{mode}} {{role}} {{state}}
```

where:

``` haskell
executable is: argv[0]
name is: [name] | <empty>
config is : config file path | <empty>
[Note: the name and config policy is (name xor config) - If a name is set, the config path is absent.]
mode is: cluster | sentinel | <empty> (empty if regular master/replica or master-with-no-replicas setup)
role is: master | replica | <empty> (empty if standalone master with no replicas)
state is: live cluster node description | live replica description | live sentinel description | <empty> (empty if no replica _or_ if a forked background process)
```

You can see example output based on different scenarios at https://gist.github.com/mattsta/7512d430b3cc43bd669e

Creation History:
- In commit 6356cf6808be9ea88ab97be33ebf1576eeb57da6, Redis changed the way it shows up in process lists.  Previously, the format was the launch command line.  After commit 6356cf6808be9ea88ab97be33ebf1576eeb57da6, then format became "redis-server ip:port" — which is _largely_ not useful in many scenarios!
  - personal note: When I have 10 Redis processes running on my server, I like to see them by logical name (typically the name of the config file shows is the name of the server), but having 10 servers with only port identifiers makes me consult _another_ lookup table to figure out what things are doing (or grep through pid files then match pid file contents to pid file names... it's not pretty).
- We've had people complaining about the change for a while.  See https://github.com/antirez/redis/issues/694 and https://github.com/antirez/redis/issues/1979
- So, after first suggesting this [back in April](https://github.com/antirez/redis/issues/694#issuecomment-41324583), I finally got around to adding a dynamic updating proctitle interface.

Implementation History:
- I started out by allowing a user-specified format string (double curly bracket formatting).  But, the format parsing code was [too ugly](https://github.com/mattsta/redis/blob/proctitle/src/redis.c#L3668-L3722), so I scrapped it in favor of [static content positions](https://github.com/mattsta/redis/blob/proctitle-noformat/src/redis.c#L3684-L3712).
- I added (or refactored out) unique state information reporters for each major mode of Redis ([master/replica](https://github.com/mattsta/redis/blob/proctitle-noformat/src/replication.c#L1514-L1526), [sentinel](https://github.com/mattsta/redis/blob/proctitle-noformat/src/sentinel.c#L2919-L2967), [cluster](https://github.com/mattsta/redis/blob/proctitle-noformat/src/cluster.c#L3560-L3607)).  I tried to select the best information to show for each mode, but we can always add or remove information as necessary.  (For example, I included cluster epoch in the cluster instance output, but not really sure why... it just looks neat.)
- The dynamic title updating is [triggered in serverCron()](https://github.com/mattsta/redis/blob/proctitle-noformat/src/redis.c#L1251-L1256) meaning:
  - to update the process title, functions call `PROCTITLE_UPDATE()` which just sets `server.update_proctitle = 1`
  - the server title update can lag server state by a second or two
  - if a process updates state multiple times, the state will only be read at most once per second (this prevents Redis from flapping the title unnecessarily).
  - Cluster and Sentinel both write out all state changes to config files — and it was very simple to just add the process title update flag to one place in the config writing code to trigger title updates too (since we know the state changed when the config updater is called).
  - Replication state tracking is a bit more tricky; it required adding `PROCTITLE_UPDATE()` triggers in a few different places since there's no centralized "replication state updated" hook.
  - Sentinel state tracking required some additional non-config-file-update-based triggers because internal per-node state (especially transitional states) aren't written out to the state/config files.
- I added a [very simple "server instance name" config option](https://github.com/mattsta/redis/commit/d9daeb778a1457867c39efd25e9d66eb894b6002) before people [started complaining about it](https://groups.google.com/forum/#!searchin/redis-db/instance$20name/redis-db/kEDe3IixFoo/moMjEOF6VFgJ).  We can either keep it, remove it, or refactor it into something more extensible.  I'd like to go further and have arbitrary "config metadata" per-instance, but that's out of the scope of what's needed for a simple process title name.

So, here it is.

If you've got free time and want to help double check everything here: try it out, try to make it crash, try to find any remaining edge cases where state updates don't get reflected in the title (meaning we need more `PROCTITLE_UPDATE()` calls somewhere), and try to think of any pathological conditions where the title updates could cause problems.

:shipit:

antirez commented:
About this: https://groups.google.com/forum/#!msg/osv-dev/3ouRUZtGSf0/4J7Ws7mKnG4J

Apparently it's good if it's possible to disable this at all, should be trivial but we apparently need an option for that.

antirez commented:
From the thread above:

---

The thing is, the setproctitle() code is already conditionally compiled, to compile only on Linux, but it doesn't help us because OSv convincingly pretends to be Linux :-)

We're already fine with setproctitle()'s assumption that argv is contiguous (OSv already supports that, see commit d511e6c091b6de11aa3bb6b16a4c41dbbe2e56fc). The only thing that breaks in OSv is the assumption that environ is also contiguous right after the end of argv. If understand correctly, setproctitle() really has no need to make this assumption! It could figure out the end of argv as argv[argc-1]+strlen(argv[argc-1]), and it shouldn't use "environ" for that (setproctitle() has nothing to do with the environment).

---

mattsta commented:
Updated by:
- Re-based against current (fixed a merge conflict)
- added config option to explicitly disable proctitle rewriting.

### Sample output

With (default) `proctitle-dynamic yes`:

``` haskell
10:41AM   0:00.02 ./redis-server *:9993 master (1 replica)
10:41AM   0:00.03 ./redis-server /Users/matt/repos/redis/redis.conf *:6379 replica (connected to 127.0.0.1:9993)
```

With the replica set to `proctitle-dynamic no`:

``` haskell
10:42AM   0:00.01 ./redis-server ../redis.conf --proctitle-dynamic no
10:41AM   0:00.03 ./redis-server *:9993 master (1 replica)
```

If you already have this checked out, you can update with `git pull -r` to have git not complain about the rebase+fix.

mattsta commented:
[fixed merge conflict caused by pid fixes in the same area of startup code (e6d499a446e1d34ceb9b393ddd79d38877aede67).]

yossigo commented:
A similar solution provided by #8397
